### PR TITLE
feature/mailserver

### PR DESCRIPTION
**Pull request by decanus**

### CODE REVIEW

File: waku/protocol/v1/waku_mail.nim [hunk 1]
```diff
@@ -76,8 +76,11 @@ proc requestMail*(node: EthereumNode, peerId: NodeId, request: MailRequest,
     return result
 
 proc p2pRequestHandler(peer: Peer, envelope: Envelope) =
-  # Mail server p2p request implementation
-  discard
+    let decoded = decode(envelope.data)
```
Comment (kdeme) [inline waku/protocol/v1/waku_mail.nim:79]: This is allowed, but if IIRC the current status-go mailservers encrypt the data with a (hardcoded??) symmetric key. 
So fine for testing I guess, but you'll have to add the option to pass a key to decrypt with. (The decode call allows passing symmetric or asymmetric key: https://github.com/status-im/nim-eth/blob/master/eth/p2p/rlpx_protocols/whisper/whisper_types.nim#L317)

File: waku/protocol/v1/waku_mail.nim [hunk 2]
```diff
@@ -76,8 +76,11 @@ proc requestMail*(node: EthereumNode, peerId: NodeId, request: MailRequest,
...
+    require decoded.isSome()
```
Comment (kdeme) [inline waku/protocol/v1/waku_mail.nim:80]: require is only for testing, in order to immediately stop the tests on failure: https://nim-lang.org/docs/unittest.html#require.t%2Cuntyped . You have to properly check `if decoded.isSome`, to continue with its data.

(And in fact, require it is also broken for usage in tests and shouldn't be used at all, I see I forgot to remove its usage in nim-waku tests)

File: waku/protocol/v1/waku_protocol.nim [hunk 3]
```diff
@@ -85,13 +85,11 @@ type
...
     accounting*: Accounting
 
-  P2PRequestHandler* = proc(peer: Peer, envelope: Envelope) {.gcsafe.}
-
   WakuNetwork = ref object
     queue*: ref Queue
     filters*: Filters
     config*: WakuConfig
-    p2pRequestHandler*: P2PRequestHandler
+    mailserver*: MailServer
```
Comment (kdeme) [inline waku/protocol/v1/waku_protocol.nim:92]: I don't think I like the current structure much. Mailserver should not become part of the waku_protocol like this.
The registration of the request handler allows you to split these off. Unless there are other issues that I currently am not seeing?

Imo the waku_mail.nim (or waku_mail_server if you like an extra file) should use (read: import) waku_protocol and not the other way around.

Comment (kdeme) [reply waku/protocol/v1/waku_protocol.nim:92]: I've done a very small refactor to highlight this better: https://github.com/status-im/nim-waku/pull/34

File: waku/protocol/v1/waku_mail.nim [hunk 4]
```diff
@@ -9,22 +9,13 @@
...
+  ./mailserver,
+  db_sqlite, ./waku_protocol
 
 const
   requestCompleteTimeout = chronos.seconds(5)
 
-type
-  Cursor = seq[byte]
-
-  MailRequest* = object
```
Comment (kdeme) [inline waku/protocol/v1/waku_mail.nim:21]: If we use a `waku_mail` and `waku_mail_server` file, then I'd rather keep these in the `waku_mail` ones. Makes more sense I think to have a simple waku client to not have to import waku_mail_server.

**Comment by oskarth** [comment]
Can we either merge this as WIP support with a follow up issue for Stimbus, or close this PR?

**Comment by oskarth** [comment]
Ping @decanus

**Comment by decanus** [comment]
https://github.com/status-im/nim-waku/issues/75

**Pull request closed without merging**